In [4]:
#Importing Liabraries
import pandas as pd
import numpy as np
import os

In [5]:
#Assigning Path

In [6]:
path=r'C:\Users\ANGEL\Desktop\Instacart Master Analysis'

In [7]:
path

'C:\\Users\\ANGEL\\Desktop\\Instacart Master Analysis'

In [8]:
#Importing Orders Data from Prepared Data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [9]:
#Importing Products Data from Original Data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [10]:
#create a dataframe
df_test = pd.DataFrame()

In [11]:
# Create a Mixed type column
df_test['mix'] = ['a','b',1,True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [14]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
df_prods.shape

(49693, 5)

In [18]:
# Creatinf a dew clean dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

In [19]:
df_prods_clean.shape

(49677, 5)

In [20]:
#Finding duplicates


In [21]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [22]:
df_prods_clean.shape

(49677, 5)

In [23]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [24]:
df_prods_clean_no_dups.shape

(49672, 5)

In [25]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# TASK 4.5

In [26]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

Task 4.5 Q2: If we look at the minimum value, the result is zero for all the columns.

In [27]:
# Check for mixed types in df_ords dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

No mixed type data coulm foun in df_ords data frame.

In [28]:
#Q5 : Finding missing values in df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [29]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [30]:
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

From the above result we can see that the only column with missing values is days_since_prior_order. Since the amount of missing values is quite large, i would not delete them. I would like to create a new column to flag these missing values.
Maybe this shows that these are all first time customers who are yet to place a second order.

In [31]:
# Addressing the Missing values
df_ords.shape

(3421083, 8)

In [32]:
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [33]:
df_ords_clean.shape

(3214874, 8)

We see that this new data frame has exactly 206209 rows less than the df_ords dataframe. Thus in the new clean df_ords_clean dataframe we have dropped these missing values.

In [34]:
# Finding duplicates in the df_ords dataframe
df_ords_clean_dups = df_ords_clean[df_ords_clean.duplicated()]


In [35]:
df_ords_clean_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [36]:
df_ords_clean_dups.size

0

These are no duplicates the the new cleaned dataframe, so we will transport the df_ords_clean file as the cheched file


In [37]:
#exporting df_ords_clean to prepared data folder
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked_1.csv'))